### PR TITLE
[FLINK-3061] Properly fail Kafka Consumer if broker is not available

### PR DESCRIPTION
**Pull request by rmetzger**

### CODE REVIEW

**Comment by aljoscha** [comment]
Could it happen that a broker is currently not available but will be in the future. With the behaviour before this PR could Flink then idle and start consuming once the broker becomes available?

File: flink-streaming-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaITCase.java [hunk 1]
```diff
@@ -40,6 +40,11 @@ public void testCheckpointing() throws Exception {
 		runCheckpointingTest();
 	}
 
+	@Test(expected = RuntimeException.class)
```
Comment (tillrohrmann) [inline flink-streaming-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaITCase.java]: You don't check the exception here. This means any `RuntimeException` can make the test pass. IMO not a good test.

**Comment by tillrohrmann** [comment]
I had a comment concerning the test case. Besides that, LGTM.

**Comment by rmetzger** [comment]
> Could it happen that a broker is currently not available but will be in the future.

Yes. With this PR, the consumer fails only if NO broker is available. As long as we can get the partition id's of the topic, the consumer will start.

> With the behaviour before this PR could Flink then idle and start consuming once the broker becomes available?

No. The parallel consuming thread deployed to the cluster are initialized with a list of partition IDs from the topic. If we can not get the partition IDs when creating the consumer, we can never assign the partitions to the consumers.

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java [hunk 2]
```diff
@@ -306,6 +306,11 @@ public FlinkKafkaConsumer(String topic, KeyedDeserializationSchema<T> deserializ
 		// Connect to a broker to get the partitions
 		List<PartitionInfo> partitionInfos = getPartitionsForTopic(topic, props);
 
+		if(partitionInfos.size() == 0) {
```
Comment (StephanEwen) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java]: Code style, space ;-)

**Comment by StephanEwen** [comment]
I think this looks good.

It only blocks job deployment if all brokers are down. As soon as there is one broker during deployment, the rest will keep recovering and retrying, if I understand it correctly.

**Comment by rmetzger** [comment]
will merge this ...

**Pull request merged**